docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,214 @@
+
+Other than integrating with CAPZ's `infrastructure-components.yaml`, the `capz-controller-manager` container itself could assume responsibility for installing ASO by using the Helm Go library to install the ASO chart when it starts. Installing ASO this way allows CAPZ to hook in additional logic, such as to check if ASO is already installed in the cluster or to verify ASO is healthy after it's installed, which would not be possible if a static manifest was embedded in `infrastructure-components.yaml`.
+
+In terms of specific options, ASO offers both single- and multi-operator configurations. The single-operator option is the default, and assumes a single ASO installation managing resources for the whole cluster. The multi-operator method allows several ASO installations to coexist, each monitoring resources in a unique set of namespaces to enable multitenancy. ASO's single-operator configuration also allows multitenancy with more tightly-scoped credentials. Overall, a single ASO installation will likely be the easier path forward since it's consistent with CAPZ's multitenancy model and doesn't involve managing ASO's CRDs and webhooks across multiple installations.
+
+#### Caveats
+
+CAPZ interacts with some Azure services that do not represent infrastructure, and thus cannot be represented in ASO. Resource Health, for example, is "reconciled" by CAPZ currently by getting a resource's health status and reflecting that in the corresponding CAPZ resource, but does not create or update any distinct Azure resources. The new SDK could be used to implement this existing functionality without affecting other service interfaces' use of ASO.
+
+Also, use of the `clusterctl move` command will require extra manual steps to move ASO resources as documented here: https://azure.github.io/azure-service-operator/introduction/frequently-asked-questions/#what-is-the-best-practice-for-transferring-aso-resources-from-one-cluster-to-another.
+
+### Security Model
+
+ASO requires credentials to interact with Azure which are documented here: https://azure.github.io/azure-service-operator/introduction/authentication/. Specifically, ASO needs the values of `$AZURE_CLIENT_ID`, `$AZURE_CLIENT_SECRET`, `$AZURE_SUBSCRIPTION_ID`, and `$AZURE_TENANT_ID` for Service Principal authentication. These credentials must be defined in a Kubernetes Secret that either applies to all ASO resources in all namespaces, all ASO resources in a particular namespace, or individual ASO resources referring to the Secret in its `serviceoperator.azure.com/credential-from` annotation. ASO allows different credentials to be used for different resources to enable multitenancy, similar to CAPZ and its AzureClusterIdentity. To mediate between the two mechanisms, CAPZ will ensure a cluster's AzureClusterIdentity or default parameters from environment variables are reflected in a Secret to be used by ASO and ensure each ASO resource it manages is created with the annotation pointing to the Secret.
```

Comment:
CAPZ does support other methods of authentication described here: https://capz.sigs.k8s.io/topics/multitenancy.html, namely Service Principal with Certificate, and User Assigned Identity. It might be a bit tricky to support all of those with ASO (for user assigned identity we could use https://azure.github.io/azure-service-operator/introduction/authentication/#deprecated-managed-identity-aad-pod-identity but not sure how we could handle a SP with certificate). 

This is a short term problem as Workload Identity support is in progress, but it might be easier to integrate ASO auth-wise once workload identity is supported.

Reply:
If ASO doesn't support Service Principal with Certificate, is there a path for those users to upgrade? Or would we expect them to first migrate to Workload Identity before upgrading?

Reply:
that's a good question, I don't have an answer. @matthchr any chance of ASO supporting SP with certificate in ASO in order to ease migration?

Reply:
>  @matthchr any chance of ASO supporting SP with a certificate in ASO in order to ease migration?

I don't see why not. Can you file an item in ASO for this @nojnhuh?

Reply:
@matthchr Yep, done: https://github.com/Azure/azure-service-operator/issues/2761